ramips: lzma-loader: use virtual memory segments for uart base address

mcprat opened the pull request:
The native bus address for UART was entered for rt305x UART_BASE, but the bootloaders have memory space remapped with the same virtual memory map the kernel uses for program addressing at boot time.

In UBoot, the remapped address is often defined as TEXT_BASE. In the kernel, for rt305x this remapped address is RT305X_SYSC_BASE. (arch/mips/include/asm/mach-ralink/rt305x.h)

Because the ralink I/O busses begin at a low address of 0x10000000, they are remapped using KSEG0 or KSEG1, which for all 32-bit MIPS SOCs (arch/mips/include/asm/addrspace.h)
are offsets of 0x80000000 and 0xa0000000 respectively. This is consistent with the other UART_BASE macros here and with MIPS memory map documentation.

Before the recent rework of the lzma-loader for ramips, the original board-$(PLATFORM).c files also did not use KSEG1ADDR for UART_BASE despite being defined, which made this mistake easier to occur.

Fix this by defining KSEG1ADDR again and actually use it. Copy and paste from the kernel's macros for consistency.

Link: https://training.mips.com/basic_mips/PDF/Memory_Map.pdf

mcprat committed:
ramips: lzma-loader: use virtual memory segments for uart base address

The native bus address for UART was entered for rt305x UART_BASE,
but the bootloaders have memory space remapped with the same
virtual memory map the kernel uses for program addressing at boot time.

In UBoot, the remapped address is often defined as TEXT_BASE.
In the kernel, for rt305x this remapped address is RT305X_SYSC_BASE.
(arch/mips/include/asm/mach-ralink/rt305x.h)

Because the ralink I/O busses begin at a low address of 0x10000000,
they are remapped using KSEG0 or KSEG1, which for all 32-bit MIPS SOCs
(arch/mips/include/asm/addrspace.h)
are offsets of 0x80000000 and 0xa0000000 respectively.
This is consistent with the other UART_BASE macros here
and with MIPS memory map documentation.

Before the recent rework of the lzma-loader for ramips,
the original board-$(PLATFORM).c files also did not
use KSEG1ADDR for UART_BASE despite being defined,
which made this mistake easier to occur.

Fix this by defining KSEG1ADDR again and actually use it.
Copy and paste from the kernel's macros for consistency.

Link: https://training.mips.com/basic_mips/PDF/Memory_Map.pdf
Fixes: c31319b66 ("ramips: lzma-loader: Refactor loader")
Reported-by: Lech Perczak <lech.perczak@gmail.com>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

Leo-PL commented:
Just tested on ZTE MF283+ - the result is strange. The board booted successfully, but the lzma-loader prompt was missing. Here is console output from booting initramfs:
```
RT3352 # setenv serverip 10.0.0.24
RT3352 # setenv ipaddr 10.0.0.254
RT3352 # tftpboot 8a100000 openwrt-ramips-rt305x-zte_mf283plus-initramfs-kernel.bin

 netboot_common, argc= 3 

 NetTxPacket = 0x83FE4E40 

 KSEG1ADDR(NetTxPacket) = 0xA3FE4E40 

 NetLoop,call eth_halt ! 

 NetLoop,call eth_init ! 
Trying Eth0 (10/100-M)

 Waitting for RX_DMA_BUSY status Start... done


 Header Payload scatter function is Disable !! 

 ETH_STATE_ACTIVE!! 
Using Eth0 (10/100-M) device
TFTP from server 10.0.0.24; our IP address is 10.0.0.254
Filename 'openwrt-ramips-rt305x-zte_mf283plus-initramfs-kernel.bin'.

 TIMEOUT_COUNT=10,Load address: 0x8a100000
Loading: Got ARP REPLY, set server/gtwy eth addr (d4:be:d9:ff:19:b7)
Got it
Got ARP REQUEST, return our IP
T T T T #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         ###########################################################
done
Bytes transferred = 4624027 (468e9b hex)
NetBootFileXferSize= 00468e9b
RT3352 # bootm 8a100000
## Booting image at 8a100000 ...
   Image Name:   MIPS OpenWrt Linux-5.15.141
   Created:      2023-12-06  12:38:37 UTC
   Image Type:   MIPS Linux Kernel Image (uncompressed)
   Data Size:    4623963 Bytes =  4.4 MB
   Load Address: 80000000
   Entry Point:  80000000
   Verifying Checksum ... OK
OK
No initrd
## Transferring control to Linux (at address 80000000) ...
## Giving linux memsize in MB, 64

Starting kernel ...

[    0.000000] Linux version 5.15.141 (builder@buildserver) (mipsel-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r0-62277c4) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0
[    0.000000] SoC Type: Ralink RT3352 id:1 rev:6
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001964c (MIPS 24KEc)
[    0.000000] MIPS: machine is ZTE MF283+
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 16kB, 4-way, VIPT, no aliases, linesize 32 bytes
```

mpratt14 commented:
apparently, there are two UARTs, and they're the same, except one is disabled by default and the other enabled by default with muxed pinout and the default for console output

but the lzma-loader is using the first one currently for rt305x (0x500)

```
	aliases {
		spi0 = &spi0;
		spi1 = &spi1;
		serial0 = &uartlite;
	};

...
...

		uart: uart@500 {
			compatible = "ralink,rt3352-uart", "ralink,rt2880-uart", "ns16550a";
			reg = <0x500 0x100>;

			clocks = <&sysc 6>;

			resets = <&sysc 12>;

			interrupt-parent = <&intc>;
			interrupts = <5>;

			reg-shift = <2>;

			status = "disabled";
		};

...
...

		uartlite: uartlite@c00 {
			compatible = "ralink,rt3352-uart", "ralink,rt2880-uart", "ns16550a";
			reg = <0xc00 0x100>;

			clocks = <&sysc 11>;

			resets = <&sysc 19>;

			interrupt-parent = <&intc>;
			interrupts = <12>;

			reg-shift = <2>;

			pinctrl-names = "default";
			pinctrl-0 = <&uartlite_pins>;
		};

```

mcprat committed:
ramips: lzma-loader: use proper register names

Before this was reworked, in the file for mt7621 subtarget
(target/linux/ramips/image/lzma-loader/src/board-mt7621.c)
the "Transmitter shift register empty" bit TEMT was used instead of
the "Transmitter holding register empty" bit THRE,
but after the rework, this value was labeled as the THRE bit instead.

Functionally there is no difference, but this is confusing to read,
as it suggests that the subtargets have different bits for the same
register in UART when in reality they are exactly the same.

One can use either bit, or both, at user's descretion
in order to determine whether the UART TX buffer is ready.
The generic kernel early-printk uses both,
(arch/mips/kernel/early_printk_8250.c)
while the ralink-specific early-printk uses only THRE,
(arch/mips/ralink/early_printk.c).

Define both bits and rewrite macros for readability,
keep the same values, as changing which to use should be tested first.

Ref: c31319b66 ("ramips: lzma-loader: Refactor loader")
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
ramips: lzma-loader: use default uart for rt305x

The rt305x series SOC have two UART devices,
and the one at bus address 0x500 is disabled by default.
Some boards do not even have a pinout for the first one,
so use the same one that the kernel uses at 0xc00 instead.

This allows the lzma-loader printing to be visible
alongside the kernel log in the same console.

Tested-by: Lech Perczak <lech.perczak@gmail.com> # zte,mf283plus
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@Leo-PL try this

Also, I'm curious if it was possible to see the "decompressing" message on the other UART if your board has it with only the first commit. It would have to be enabled in DTS.

but you don't have to waste time with that if you don't want to :P

Leo-PL commented:
> @Leo-PL try this
> 
> Also, I'm curious if it was possible to see the "decompressing" message on the other UART if your board has it with only the first commit.
> 
> but you don't have to waste time with that if you don't want to :P

Not sure if there are even any testpoints on this one for the 2nd UART. I'll take a look later today or tomorrow.

Leo-PL commented:
> @Leo-PL try this
> 
> Also, I'm curious if it was possible to see the "decompressing" message on the other UART if your board has it with only the first commit.
> 
> but you don't have to waste time with that if you don't want to :P

Just took a look at the board, but I couldn't locate any testpoints that could lead to 2nd UART. But with the latest patches applied, board boots fine and the lzma-loader prompt is visible. Great work! And thank you for prompt response!

Feel free to add `Tested-by: Lech Perczak <lech.perczak@gmail.com>`

While we're at it, let's notify @hauke, as this will require backport to 23.05 branch, too - all 23.05.x images for MF283+ are broken - and possibly for other rt305x devices too.

There are some fixes to be done for `mtk_eth_soc` for them, otherwise kernel can sometimes panic during boot if traffic comes in a wrong moment - but let's cover that in a separate issue. I'll handle that in the morning.

mpratt14 commented:
> I couldn't locate any testpoints that could lead to 2nd UART

ohhh... I thought you were talking about "points in commit history" or something like that... haha

I updated commit message to reflect the fact that some boards don't even have it.


> thank you for prompt response!

you're welcome, you happened to catch my attention at the right time 😆

Leo-PL commented:
I cherry-picked the changes to 23.05 branch for my local build - they obviously work perfectly, too.

mpratt14 commented:
ping @hauke

ready for review

xabolcs commented:
Fixes half of #14138

hauke commented:
Please create a pull request with these changes for 23.05.

Leo-PL commented:
@hauke does it make sense if I combine these changes in single PR with my Ethernet fixes?
